[quant][graphmode] Support quantizing any methods called

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25974 [quant][graphmode] use whitelist for selecting observed values
* #25625 [quant][graphmode] Fold quantize op into module
* #25624 [quant][graphmode] Add fusion for quantized linear
* #25623 Port fuse_linear from pytorch/tvm
* **#25505 [quant][graphmode] Support quantizing any methods called**
* #25504 [quant][graphmode] Skip inserting duplicate observers

Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)

jerryzh168 committed:
[quant][graphmode] Support quantizing any methods called

Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -575,51 +575,48 @@ void InsertQuantDeQuantImpl(
     Block* b = blocks_to_visit.top();
     blocks_to_visit.pop();
-    for (Node* n : b->nodes()) {
+    for (auto it = b->nodes().begin(), end=b->nodes().end(); it != end;) {
```
Nit: clang-format :-p

jerryzh168 replied:
can we make it a command hook or something? do you have the script?

ZolotukhinM replied:
I use `git clang-format` - it formats not-committed changes. You could add it as a pre-commit hook, but I don't use it because it messes up reverts if the original version was ill-formatted.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -575,51 +575,48 @@ void InsertQuantDeQuantImpl(
+            if (module_instance == graph->inputs()[0]) {
+              InsertQuantDeQuantImpl(module, module_method_name);
+            } else {
+              auto child_module = qh.findChildModuleToQuantize(v);
```
I think this place can be cleaned up a bit.

For starters, we don't need `if (v->node()->kind() == prim::CallMethod) {` check in `findChildModuleToQuantize` anymore.

Next, I think the code above is very similar to the code inside that function, so I think it would be good if we can just reuse it. The only difference is that in one case we check `module_instance == graph->inputs()[0]` and in the other we're looking at submodules. If that's refactoring doesn't work out, I think it would still make sense to factor out the code above to a separate function, so that we're operating on the same level of details in a scope of one function (right now the code above is one level lower than a call to `findChildModuleToQuantize` which follow that code).

jerryzh168 replied:
Yeah I think there are some duplicated code to get module_instance and module_method_name, I can remove these in findChildModuleToQuantize function

jerryzh168 replied:
I think in both cases we need to know module_instance and module_method_name, so it's not different level of details, just in the case of child modules we also need to get the child module, and in case of self, the module is already there so we don't need to call another function to get it.

ZolotukhinM commented on test/test_quantizer.py:
```diff
@@ -85,6 +83,7 @@ def get_forward(m):
             return m._c._get_method('forward')
         # TODO: test jit.script as well
         torch._C._jit_pass_constant_propagation(get_forward(script_module).graph)
+        torch._C._jit_pass_constant_propagation(script_module._c._get_module('conv')._get_method('conv2d_forward').graph)
```
I'm still not sure why we need to run constant prop.

jerryzh168 replied:
this is just to make checking easier, if we don't do this, we'll see many branches and many observers. Also I haven't test the code with branches yet, so it might break in these cases.

ZolotukhinM replied:
Ok, let's keep it for now.

ZolotukhinM reviewed (approved):
Overall looks good, please find some comments inline.

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)

nishantpdce commented:
LGTM

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -514,51 +511,48 @@ void InsertQuantDeQuantImpl(
+            } else {
+              auto child_module = qh.findChildModuleToQuantize(module_instance);
+              if (child_module) {
+                InsertQuantDeQuantImpl(child_module.value(), module_method_name);
```
Nit: might be worth moving `InsertQuantDeQuantImpl` after the if so that there is only one place where we invoke it. Then the code before it can be (optionally) factored out to a separate function.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -514,51 +511,48 @@ void InsertQuantDeQuantImpl(
-          if (child_module) {
-            child_modules_to_quantize[child_module.value().module_object()] =
-                child_module.value();
+          if (v->node()->kind() == prim::CallMethod) {
```
Nit: The if above can be changed to `if (!isTensor) { continue; ... }` so that we have less `}` braces in the code afterwards (it makes it easier to see what `if` an `else` relates to and in general makes the code more stream-lined).

ZolotukhinM reviewed (approved):
Still looks good :)

Nit: the commit message mentions that it needs the const prop but I don't see anything in the code requiring that. How does it depend on it?

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

It relies on module level constant prop, we need to figure out a way to do constant prop
for these methods as well. We can either do constant prop in the module level or do constant
prop in the quantization function, but this will need some discussion.

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing any methods called"


Summary:
Support for quantizing all the methods called by forward method, including
child module methods and other methods in the current module

Test Plan:
python test/test_jit.py 'TestJit.insert_quant_dequant'
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208887](https://our.internmc.facebook.com/intern/diff/D17208887)
